Step2 : 문자열 계산기

hbm0411 opened the pull request:
문자열 계산기 구현 완료하였습니다.

* main 함수에 대한 입출력 테스트도 가능할까요?

hbm0411 committed:
feat: Readme에 요구 사항 추가

hbm0411 committed:
feat: calculator main 클래스 추가

hbm0411 committed:
feat: 사칙연산 enum class 및 관련 테스트 추가

hbm0411 committed:
feat: 입력받은 문자열 validate 및 리스트로 분리

hbm0411 committed:
refactor: 잘못된 연산식에 대한 validation 추가

hbm0411 committed:
feat: 계산 로직 및 테스트 추가

hbm0411 commented on src/main/kotlin/calculator/Operator.kt:
```diff
@@ -0,0 +1,25 @@
+
+import calculator.Const.ErrorMsg
+
+// TODO interface vs 람다식 중에 어떤게 더 좋을까? https://kotlinlang.org/docs/enum-classes.html#implementing-interfaces-in-enum-classes
```
https://kotlinlang.org/docs/enum-classes.html#implementing-interfaces-in-enum-classes
를 보면 이런 예시가 있는데요, 일단 좀 더 간단한 람다식으로 처리했습니다.
저렇게 implement 하여 구현하는 것에 대한 장점이 있을까요?

```
enum class IntArithmetics : BinaryOperator<Int>, IntBinaryOperator {
    PLUS {
        override fun apply(t: Int, u: Int): Int = t + u
    },
    TIMES {
        override fun apply(t: Int, u: Int): Int = t * u
    };

    override fun applyAsInt(t: Int, u: Int) = apply(t, u)
}
```

laco-dev replied:
람다를 프로퍼티로 가지는 것과 인터페이스를 구현하는 것은 명확한 역할의 차이가 있다고 생각이 드는데요.

장점으로 바라본다면
1. 이름이 있는 인터페이스를 제공한다
2. 두 개 이상의 인터페이스를 구현할 수 있다
3. 두 개 이상의 인터페이스 함수를 만들 수 있다
4. ...

지금 구현에서는 한개의 인터페이스 메소드만 구현하면 되기 때문에 가독성을 비롯해 람다를 사용하는게 보다 좋아 보입니다.

laco-dev commented on README.md:
```diff
@@ -1 +1,15 @@
+### 상세 구현 사항
+- [ ] 입력값이 null이거나 빈 공백 문자일 경우 IllegalArgumentException throw
+- [ ] 사칙연산 기호가 아닌 경우 IllegalArgumentException throw
+- [ ] 사칙 연산을 모두 포함하는 기능 구현
```
구체적인 요구사항 정리 너무 좋습니다. 👍 
조금 더 작은 단위로 기능을 나눠봐도 좋을 것 같아요!

- [ ] 더하기 기능 구현
- [ ] 빼기 기능 구현
- [ ] 곱하기 기능 구현
- [ ] 나누기 기능 구현
- [ ] ...

hbm0411 replied:
넵 알겠습니다.

laco-dev commented on src/test/kotlin/calculator/OperatorTests.kt:
```diff
@@ -0,0 +1,64 @@
+
+
+class OperatorTests {
+    @Nested
+    @DisplayName("사칙연산 계산 테스트")
+    inner class OperatorExpression {
```
Nested 테스트를 활용해 테스트 가독성을 올리는 개념이군요! 👍

hbm0411 replied:
넵 맞습니다. 테스트가 많아지면 비슷한 테스트끼리 그룹화해서 묶어주는게 가독성이 좋더라구요 ㅎㅎ

laco-dev commented on src/test/kotlin/calculator/OperatorTests.kt:
```diff
@@ -0,0 +1,64 @@
+    inner class OperatorExpression {
+        @Test
+        fun `덧셈 성공`() {
+            assertThat(Operator.PLUS.expression(3.0, 4.0)).isEqualTo(7.0)
+        }
+
+        @Test
+        fun `뺄셈 성공`() {
+            assertThat(Operator.MINUS.expression(10.0, 3.6)).isEqualTo(6.4)
+        }
+
+        @Test
+        fun `곱셈 성공`() {
+            assertThat(Operator.TIMES.expression(10.0, 3.6)).isEqualTo(36.0)
+        }
+
+        @Test
+        fun `나눗셈 성공`() {
+            assertThat(Operator.DIVIDE.expression(10.0, 2.0)).isEqualTo(5.0)
+        }
```
테스트 이름에 덧셈, 뺄셈, 곱셈, 나눗셈 이라는 동작이 우리가 생각하기에 당연한 결과를 기대할 수 있는데요.
한가지 아쉬운 점은 로직과 로직의 결과를 머릿속에서 한번 더 계산해야 하는 문제가 있습니다.

예를 들어 10.0 나누기 2.0 이라는 결과가 5.0에 도달하기 위해 한번 스스로 계산을 해야 하죠.. 😢 

테스트 이름을 보다 에상값과 결과가 직관적으로 드러나도록 만들어 보면 어떨까요?
위 경우에서는 
```
10.0과 3.6을 곱하면 36.0이다 
10.0과 2.0을 나누면 5.0이다
```

hbm0411 replied:
함수 이름에 바로 `10.0과 3.6을 곱하면 36.0이다` 를 작성하다 보니
`"."` 문자열 때문에 불가능해서
부득이하게 `@DisplayName`으로 작성하였습니다.
다른 방법이 있다면 알려주시면 감사하겠습니다 :)

laco-dev commented on src/main/kotlin/calculator/Operator.kt:
```diff
@@ -1,18 +1,25 @@
+            require(values().any { it.symbol == operatorStr }) { ErrorMsg.OPERATOR_IS_NOT_VALID_ERROR_MSG }
+        }
+
+        fun find(operatorStr: String?): Operator {
+            checkValidation(operatorStr)
+            return values().first { it.symbol == operatorStr }
```
find라는 기능이 만들어지면서, checkValidation 이라는 함수는 협력하기 위한 관계(공개 인터페이스)가 아닌  내부적으로 검증하기 위한 함수로 변경이 되었는데요.

테스트를 위해서 가시성이 열려있을 필요는 없기 때문에 checkValidation은 private으로 만들어도 될 것 같습니다!

laco-dev replied:
find 함수를 통해 전체 enum values가 두번을 순회하고 있습니다.
한번만 순회하면서 의도에 맞는 오류를 발생시키게 만들어 볼 수 있을 것 같아요.
```kotlin
val operator = values().find { ... }
return requireNotNull(operator) { ... }
```

hbm0411 replied:
나중에 find 함수를 만들게 되면서 고민했던 부분인데 
역시나 하나로 합치는 방향으로 생각해야겠습니다ㅎㅎ
너무 메소드를 분리하려고 욕심부린 느낌(?) 이네요

laco-dev commented on src/main/kotlin/calculator/InputProcessor.kt:
```diff
@@ -0,0 +1,29 @@
+object InputProcessor {
+    private const val INPUT_DELIMITER = " "
+
+    fun checkInputIsEmpty(expression: String?): String {
+        require(!expression.isNullOrBlank()) { ErrorMsg.INPUT_IS_EMPTY_OR_BLANK_ERROR_MSG }
+        return expression
+    }
```
코틀린의 오류 표준함수중 check { ... } 형태는 내부적으로 IllegalStateException을 반환하고 있어서
이를 사용하는 개발자가 오류를 착가할 수 있을 것 같아요.

동일한 이름 패턴으로 지어보면 어떨까요?
```kotlin
fun requireNotNullOrBlank(...)  // <-- 이건 ArgumentException ?
fun checkNotNullOrBlank(...) // <-- 이건 StateException ?

```

hbm0411 replied:
반영했습니다 :)

laco-dev commented on src/test/kotlin/calculator/InputProcessorTests.kt:
```diff
@@ -0,0 +1,66 @@
+    @Nested
+    @DisplayName("입력값이 null 이거나 빈칸인지 테스트")
+    inner class CheckInputIsEmptyOrBlank {
+        @ParameterizedTest
+        @ValueSource(strings = ["3", "1+2", "1 + 2"])
+        fun `입력값이 Null 이나 빈칸이 아닐 경우 성공`(expression: String) {
+            assertDoesNotThrow { InputProcessor.checkInputIsEmpty(expression) }
+        }
```
성공하는 케이스에 대해 단순히 오류가 안나는 것으로 판단하기 것은 지양하는것이 좋습니다.
이 기능을 테스트 하려면 어떻게 해야 할지 고민 해 보시겠어요?

hbm0411 replied:
해당 메소드가 NotNull인 String을 반환하기 때문에 `assertThat().isNotNull` 로 수정하였습니다.

laco-dev commented on src/test/kotlin/calculator/InputProcessorTests.kt:
```diff
@@ -0,0 +1,66 @@
+            assertDoesNotThrow { InputProcessor.checkInputIsEmpty(expression) }
+        }
+
+        @ParameterizedTest
+        @NullAndEmptySource
+        @ValueSource(strings = ["  ", "\t"])
+        fun `입력값이 null 이거나 빈칸이면 실패 - IllegalArgumentException 발생`(expression: String?) {
+            assertThatIllegalArgumentException()
+                .isThrownBy { InputProcessor.checkInputIsEmpty(expression) }
+                .withMessage(ErrorMsg.INPUT_IS_EMPTY_OR_BLANK_ERROR_MSG)
+        }
```
`@ParamterizedTest`에 name으로 테스트 이름에 파라미터를 넣어서 확인할 수 있습니다.
아래와 같이 만들어 볼 수 있어요!
```kotlin
@ParamterizedTest(name = "입력값이 `{0}` 이면 IllegalArgumentException이 발생한다`
```
![image](https://user-images.githubusercontent.com/28249981/167440864-e7de04f6-15e0-4c3e-80fe-a0a5420bab64.png)

hbm0411 replied:
테스트 이름이 구체적으로 보여서 좋네요! 반영했습니다.

laco-dev commented on src/main/kotlin/calculator/InputProcessor.kt:
```diff
@@ -0,0 +1,29 @@
+                it.toDoubleOrNull() ?: throw NumberFormatException(ErrorMsg.OPERAND_IS_NOT_NUMBER_ERROR_MSG)
+            }
+
+        return Pair(operators, operands)
```
Kotlin에서 Pair를 만들때는 직접 생성자를 이용하기 보다 to 라는 중위 연산자를 활용할 수 있습니다!
```kotlin
operators to operands
```

hbm0411 replied:
오! pair를 자주 쓰는 편인데 몰랐네요. 감사합니다.

laco-dev commented on src/main/kotlin/calculator/InputProcessor.kt:
```diff
@@ -0,0 +1,29 @@
+        return expression
+    }
+
+    fun splitInputToPairList(expression: String): Pair<List<Operator>, List<Double>> {
```
함수 이름이 PairList면 무엇을 반환하는지 추측이 어려울 것 같습니다.
Operator와 Operand라는 이름을 이용해 보다 잘 드러나도록 바꿔볼 수 있을 것 같아요.

laco-dev commented on src/main/kotlin/calculator/CalculatorProcessor.kt:
```diff
@@ -0,0 +1,17 @@
+package calculator
+
+object CalculatorProcessor {
+    fun calculate(operators: List<Operator>, operands: List<Double>): Double {
```
바깥에서 InputProcessor를 이용해 검증을 다 하고 왔는데 정작 calculate 함수에서는 하고 있지 않아서 잘못된 파라미터를 넣을 가능성이 생겨버리네요. 😭 

expression을 메시지의 입력 값으로 받아 InputProcessor를 의존하게 하는 것도 괜찮을 것 같네요!

hbm0411 replied:
의존성을 줄여보려는 의도였는데 아무래도 이게 더 맞는 구조인것 같네요. 반영했습니다 :)

laco-dev commented on src/main/kotlin/calculator/CalculatorProcessor.kt:
```diff
@@ -0,0 +1,17 @@
+    fun calculate(operators: List<Operator>, operands: List<Double>): Double {
+        var result = operands.first()
+
+        println("\n========== 계산 시작 ==========")
+        operands.drop(1).zip(operators) { operand, operator ->
+            print("$result ${operator.symbol} $operand = ")
+            result = operator.expression(result, operand)
+            println(result)
+        }
+        println("========== 계산 완료 ==========\n")
```
CalculatorProcessor는 도메인 로직이기 때문에 View (print) 코드가 위치하면 안됩니다.

laco-dev commented on src/test/kotlin/calculator/CalculatorProcessorTests.kt:
```diff
@@ -0,0 +1,22 @@
+    inner class CalculatorProcessorTest {
+        @Test
+        fun `사칙 연산 계산 성공`() {
+            // 2 + 3 * 4 / 2 = 10
+            val operators = listOf(Operator.PLUS, Operator.TIMES, Operator.DIVIDE)
+            val operands = listOf(2.0, 3.0, 4.0, 2.0)
+
+            val result = CalculatorProcessor.calculate(operators, operands)
+            assertThat(result).isEqualTo(10.0)
```
CsvSource를 이용하면 여러 연산을 한번에 테스트 해 볼 수 있습니다.
```kotlin
CsvSource(
   value = [
      "2 + 3 * 4 / 2 = 10"
      "1 + 2 * 5 = 15"
   ]
```
ParamterizedTest의 name과 조합해서 테스트 케이스를 추가 해 보세요.

hbm0411 replied:
이전 구조에서는 Operator type의 리스트와 operand 리스트를 받아야 해서 csvSource를 사용하지 못했는데,
inputStr을 받도록 구조를 수정하며 CsvSource를 사용하도록 수정하였습니다.

laco-dev reviewed (changes requested):
하나라도 놓치지 않겠다는 의지로 꼼꼼하게 잘 작성 해 주셨네요! 👍 
객체지향 관점에서도 많은 경험이 있으신 것 같아요.

관련해서 고민해볼 내용과 테스트 쪽에서 몇 가지 피드백을 남겨드렸습니다!
확인 후 다시 리뷰 요청 주세요~

hbm0411 committed:
refactor: 리뷰 반영

- Pair 생성자 대신 to 중위연산자 사용하기
- Readme에 상세 구현 사항 좀 더 자세히 작성
- OperatorTests의 테스트명 직관적으로 작성
- Operator의 checkValidation() 과 find()를 합쳐 불필요한 순회 줄이기
- assertDoesNotThrow 테스트 지양
- @ParameterizedTest의 name값을 이용하여 테스트명 구체적으로 작성

hbm0411 committed:
refactor: 리뷰 반영 - CalculatorProcessor 구조 수정

- CalculatorProcessor 안에서 InputStr을 받아 valid check 하도록 수정
- CalculatorProcessor 에서 View(print) 코드 삭제
- CalculatorProcessorTests 코드에 csvSource 이용

hbm0411 committed:
refactor: klint 적용

laco-dev commented on src/test/kotlin/calculator/OperatorTests.kt:
```diff
@@ -15,21 +15,25 @@ class OperatorTests {
     @DisplayName("사칙연산 계산 테스트")
     inner class OperatorExpression {
         @Test
+        @DisplayName("덧셈 성공 - 3.0과 4.0을 더하면 7.0")
```
ㅎㅎ 이런 목적으로도 `@DisplayName`이 존재하니까요.
저는 더욱 좋아졌다고 생각이 들어요!

단순히 덧셈 성공이라는게 무슨 덧셈이지 라고 생각했던거에 비해 이제는 확실하게 결과가 들어오네요 💯

laco-dev reviewed (approved):
피드백 확인 했습니다.
의도를 잘 파악하고 멋지게 반영 해 주셨네요! 💯 
코틀린에 숨겨진 좋은 기능들이 더 많으니 같이 진행하면서 발견 해 나가기로 하시죠!

2단계 미션은 종료하겠습니다. 🎉
